### PR TITLE
Update mono to 2017-02 "monthly master" branch [do not merge yet]

### PR DESCRIPTION
**Pull request by xmcclure**
To be merged after Wrench testing is complete and after the next xamarin-android branch.

### CODE REVIEW

File: build-tools/mono-runtimes/mono-runtimes.targets [hunk 1]
```diff
@@ -8,7 +8,8 @@
   </PropertyGroup>
   <ItemGroup>
     <MonoDocCopyItem Include="monodoc.dll" />
-    <MonoDocCopyItem Include="monodoc.dll.mdb" />
+    <MonoDocCopyItem Include="monodoc.pdb"  Condition="Exists ('monodoc.pdb')" />
```
Comment (jonpryor) [inline build-tools/mono-runtimes/mono-runtimes.targets]: I don't think this `Condition` can *possibly* work. It's going to use the current working directory, and there are no odds that the current working directory will be `external/mono/mcs/class/lib/net_4_x`.

I *think* what you'd actually want to do is remove the `Condition` on this line and the following line, and instead update `@(_MonoDocCopyItems)` -- line 30? -- to be:

```xml
<ItemGroup>
  <_MonoDocCopyItems
      Condition=" Exists ('$(_MonoOutputDir)\%(Identity)') "
      Include="@(MonoDocCopyItem->'$(_MonoOutputDir)\%(Identity)')"
  />
  <_MonoDocInstalledItems
      Condition=" Exists ('$(_MonoOutputDir)\%(Identity)') "
      Include="@(MonoDocCopyItem->'$(_MandroidDir)\%(Identity)')"
  />
</ItemGroup>
```

@dellis1972: Does the above sound correct to you?

Comment (jonpryor) [inline build-tools/mono-runtimes/mono-runtimes.targets]: This won't actually work either; top-level `<ItemGroup/>` values are processed at file-load time, i.e. when `mono-runtimes.targets` is processed, and thus is quite possible that Mono won't have been built yet, and thus the `Condition` expressions here will *still* fail.

I think we'll need a "helper" target?

Comment (dellis1972) [inline build-tools/mono-runtimes/mono-runtimes.targets]: @xmcclure sorry, I completely forgot that we needed the path in there. 

Best thing to do is to write a target which will collect the files that exist. You can use the ItemGroup code that @jonpryor mentioned and pop that in the new Target. That will add the items to the ItemGroup at run time not load time. 

```
<Target Name="_CollectDebuggingFiles">
<ItemGroup>
  <_MonoDocCopyItems
      Condition=" Exists ('$(_MonoOutputDir)\%(Identity)') "
      Include="@(MonoDocCopyItem->'$(_MonoOutputDir)\%(Identity)')"
  />
  <_MonoDocInstalledItems
      Condition=" Exists ('$(_MonoOutputDir)\%(Identity)') "
      Include="@(MonoDocCopyItem->'$(_MandroidDir)\%(Identity)')"
  />
</ItemGroup>
</Target>
```

Then just add the _CollectDebuggingFiles target to the DependsOnTargets list of the target you need it to run before.

**Comment by jonpryor** [comment]
Additionally, you need to update `@(BundleItem)`, in `GetMonoBundleItems` (around line 479). If you go with the `_CollectDebuggingFiles` approach, then `_CollectDebuggingFiles` needs to be added to `DependsOnTargets` for `GetMonoBundleItems` (line 477). If you opt to treat debug symbol files separately, you'll need to separately update `@(BundleItem)`.

**Comment by xmcclure** [comment]
Should the existing `_InstallMonoSymbolicate` target depend on anything?

**Comment by xmcclure** [comment]
Okay, I think this is correct (please see full commit notes for details of what I did), and after doing a git clean xffd and rebuild the pdb appears to be getting copied to the right directory...
```andi:xamarin-android andi$ find ./* | grep pdb | grep -i monodoc
./bin/Debug/lib/mandroid/monodoc.pdb
./external/mono/mcs/class/lib/net_4_x/monodoc.pdb```
Note, I had to add a bunch of DependsOnTargets-es to make this all work right. I did not touch the Depends logic for anything not specifically related to my changes, however. I think you need to audit the dependencies of all items in ForceBuildDependsOn. Your current logic where you assume that separating X;Y by semicolons is enough to ensure X executes before Y is not safe, and (apparently!) can get subverted merely because a `DependsOnTargets` was added in a different part of the file.

File: build-tools/mono-runtimes/mono-runtimes.targets [hunk 2]
```diff
@@ -323,9 +317,27 @@
...
   </Target>
+  <Target Name="_GetMonodocItems" DependsOnTargets="_BuildRuntimes">
+    <ItemGroup>
+      <_MonoDocCopyItems Include="@(MonoDocCopyItem->'$(_MonoOutputDir)\%(Identity)')" />
+      <_MonoDocCopyItems
+          Condition=" Exists ('$(_MonoOutputDir)\%(Identity)') "
+          Include="@(MonoDocCopyItemOptional->'$(_MonoOutputDir)\%(Identity)')"
+      />
+    </ItemGroup>
+    <ItemGroup>
```
Comment (dellis1972) [inline build-tools/mono-runtimes/mono-runtimes.targets]: We can collapse these into one ItemGroup. there is not need to separate them. unless @jonpryor said to :P

Comment (xmcclure) [inline build-tools/mono-runtimes/mono-runtimes.targets]: They're not the same, one contains "mdoc.exe" and the other doesn't

Comment (dellis1972) [inline build-tools/mono-runtimes/mono-runtimes.targets]: You can mix different items within an ItemGroup so 

```xml
<ItemGroup>
    <Foo Include="Somefile.dll" />
    <Bar Include="Someotherbarfile.dll" />
</ItemGroup>
```

MSbuild and xbuild will evaluate each "item" separately. But its sometimes useful to put items together in logical groups just to make it easier to maintain. This is probably one of those cases :)

**Comment by xmcclure** [comment]
Moved to PR #445

**Pull request closed without merging**